### PR TITLE
Enhanced/automatically tab to edit

### PR DESCRIPTION
**Pull request by mohammedhemaid**
A video for it
https://photos.app.goo.gl/223pjfBTEXGt3TPG9

### CODE REVIEW

File: app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java [hunk 1]
```diff
@@ -259,6 +264,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
                 return true;
             case R.id.view_tutorial:
                 showTutorial(false);
+                UIUtils.hideKeyboard(this);
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java]: If you update this branch against Gear61:master, this should go away. It's not part of this pull request, so let's try to get it out to keep the pull request focused. At Facebook, we have a concept of "One and only thesis", which states that every pull request should be hyper-focused and only do 1 thing.

File: app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java [hunk 2]
```diff
@@ -42,6 +42,9 @@
 public class MainActivity extends StandardActivity {
 
     public static final String LIST_NAME_KEY = "listName";
+    public static final int NAME_CHOOSING_FRAGMENT = 0;
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java]: Some thoughts here:
1. I feel like the receiver (ListActivity) should contain the keys, not the sender (MainActivity). Putting the keys within the sender makes this functionality less flexible in the long run. If ListActivity can be started from another place in the future, we will need to define a new set of keys, which will lead to dangerous copy-paste. Think about this problem as ListActivity "exposing" an API for other places to call to control its UI/UX.
2. As mentioned in class, each component should "hide" as much of its functionality as possible. By passing in the ViewPager index indirectly, ListActivity is revealing that it uses a ViewPager. This is fragile in the future since ListActivity may move off of a ViewPager implementation. Let's just create a new key like "START_ON_EDIT_PAGE" and the value will be a true/false boolean.
3. Do the simple thing first! Right now, there isn't a need to expose such a powerful API since by default, the ViewPager starts off on index 0 and we only need to make it 1 every once in a while. A powerful API also comes with risk, since a caller might pass in 2 or something and cause index out of bounds.

File: app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java [hunk 3]
```diff
@@ -166,6 +169,7 @@ public void onItemClick(int position) {
         Intent intent = new Intent(this, ListActivity.class);
         String listName = nameListsAdapter.getItem(position);
         intent.putExtra(LIST_NAME_KEY, listName);
+        intent.putExtra(LIST_TYPE, NAME_CHOOSING_FRAGMENT);
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java]: Once we move to a simple, declarative boolean passing model, we won't need this line. 👍

**Review by Gear61 — CHANGES_REQUESTED** [review]
This is a good start! Love the recorded video as the test plan (shows that your code works) and I see that you regression tested the other case as well. Hope you learn from the comments I left.

File: app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java (hunk 1 above)
Comment (mohammedhemaid) [reply app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java]: I think I did a mistake here, Should I merge the branch with the master and then continue the second issue?
Or make the last commit as ahead for the new branch and then continue from there

File: app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java [hunk 4]
```diff
@@ -174,6 +177,7 @@ public void addItem() {
             nameListsAdapter.addList(newList);
             Intent intent = new Intent(this, ListActivity.class);
             intent.putExtra(LIST_NAME_KEY, newList);
+            intent.putExtra(LIST_TYPE, START_ON_EDIT_PAGE);
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java]: Sorry for the miscommunication; I meant that the key should be:

`public static final String START_ON_EDIT_PAGE = "startOnEditPage"`

And we just pass in true as the value here.

File: app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java [hunk 5]
```diff
@@ -42,12 +46,17 @@ protected void onCreate(Bundle savedInstanceState) {
         String listName = getIntent().getStringExtra(MainActivity.LIST_NAME_KEY);
         setTitle(listName);
 
+        boolean startFormEditPage = getIntent().getBooleanExtra(LIST_TYPE,false);
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java]:
```suggestion
        boolean startFromEditPage = getIntent().getBooleanExtra(START_ON_EDIT_PAGE, false);
```

**Review by Gear61 — CHANGES_REQUESTED** [review]
Thanks for incorporating the feedback! Should be good after 1 more round of changes.

File: app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java [hunk 6]
```diff
@@ -174,6 +176,7 @@ public void addItem() {
             nameListsAdapter.addList(newList);
             Intent intent = new Intent(this, ListActivity.class);
             intent.putExtra(LIST_NAME_KEY, newList);
+            intent.putExtra(START_ON_EDIT_PAGE , true );
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/home/MainActivity.java]:
```suggestion
            intent.putExtra(START_ON_EDIT_PAGE, true);
```

File: app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java [hunk 7]
```diff
@@ -42,12 +45,17 @@ protected void onCreate(Bundle savedInstanceState) {
...
 
+        boolean startFormEditPage = getIntent().getBooleanExtra(START_ON_EDIT_PAGE,false);
+
         preferencesManager = new PreferencesManager(this);
         ListTabsAdapter listTabsAdapter = new ListTabsAdapter(
                 getSupportFragmentManager(),
                 listName,
                 listTabTitles);
         nameListPager.setAdapter(listTabsAdapter);
+        if (startFormEditPage){
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java]: nit: I think we should just inline getIntent().getBooleanExtra(START_ON_EDIT_PAGE, false) within this if statement.

File: app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java [hunk 8]
```diff
@@ -21,6 +21,9 @@
 
 public class ListActivity extends StandardActivity implements ShakeDetector.Listener {
 
+    public static final int EDIT_NAME_LIST_FRAGMENT = 1;
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java]: nit: Removing fragment from the name hides implementation, so I think this is slightly cleaner.
```suggestion
    public static final int EDIT_NAME_LIST_PAGE_POSITION = 1;
```

**Review by Gear61 — CHANGES_REQUESTED** [review]
Looks a lot better! Just a couple small changes left.

File: app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java [hunk 9]
```diff
@@ -21,6 +21,9 @@
 
 public class ListActivity extends StandardActivity implements ShakeDetector.Listener {
 
+    public static final int EDIT_NAME_LIST_PAGE_POSITION = 1;
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java]: nit: This should be private. Let's hide the implementation.
```suggestion
    private static final int EDIT_NAME_LIST_PAGE_POSITION = 1;
```

File: app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java [hunk 10]
```diff
@@ -48,6 +51,9 @@ protected void onCreate(Bundle savedInstanceState) {
                 listName,
                 listTabTitles);
         nameListPager.setAdapter(listTabsAdapter);
+        if (getIntent().getBooleanExtra(START_ON_EDIT_PAGE,false)){
```
Comment (Gear61) [inline app/src/main/java/com/randomappsinc/studentpicker/listpage/ListActivity.java]: nit: Small spacing issue
```suggestion
        if (getIntent().getBooleanExtra(START_ON_EDIT_PAGE, false)) {
```

**Review by Gear61 — CHANGES_REQUESTED** [review]
Please increase the versioning as well. Should be versionName of "3.9.11" and versionCode of 401.

**Review by Gear61 — APPROVED** [review]
Awesome! 👍
